### ISSUE TITLE
[Bug]: story rendering order influences the order of the CSS styles in the head

### ISSUE THREAD

**Issue by martinkutter** [storybookjs/storybook#23990]
### Describe the bug

As discovered some years ago in #13515 the name of the story defines in which order the imported styles from the component are added to the head. So if I change the name of a story, or delete the story, or force a rerender the look of a component could change.

### To Reproduce

https://stackblitz.com/edit/github-64tjnp

<img width="166" alt="image" src="https://github.com/storybookjs/storybook/assets/16589094/83c922ab-625d-42e2-92c9-d47975c6237d">

`AHeader` imports `header.module.scss` so it is added first in the head.
`ZHeader` imports `header.module.scss` again, but it's already in the head.
As a result, the button turns green because `button.modules.scss` is added to the head after `header.module.scss`.
<img width="523" alt="image" src="https://github.com/storybookjs/storybook/assets/16589094/8cffa90d-ca0c-41bd-9f34-b42271f842f9">

If I simply resize the window, the order changes and the button turns red.
![Kapture 2023-08-29 at 15 53 09](https://github.com/storybookjs/storybook/assets/16589094/6c52ed57-289d-4b7d-aa67-28b9239f8bfa)

**Comment by dcop** [comment]
This is happening to me also on 7.5.1, but the weird thing is that in dev mode everything works fine, while building and serving is breaking the CSS order. I'm also using CSS modules and my main (css) theme is loaded in the provider that I then use in the `preview` file with `withThemeFromJSXProvider`.

**Comment by tmeasday** [comment]
I'm not an expert on CSS modules, but in my experience, if you see different rendering behaviour when you access stories/components in a different order, you'll see the same issue in your actual code depending on the order you import those components in your source code. That import order is often driven by the order you visit routes in your app and can be super tricky to track down.

That seems to me to be a bug that you want to fix in your code and will be much easier to debug in SB 🤷 

In the example repro posted (thank you!) I can't quite get the CRA app running in StackBlitz but I downloaded it and added this to the `index.tsx` to simulate the behaviour of a router that does dynamic imports (e.g like NextJS):

```ts
import React from 'react';
import ReactDOM from 'react-dom/client';

const ZHeader = React.lazy(async () => ({ default: (await import('./stories/zHeader')).ZHeader }));
const AHeader = React.lazy(async () => ({ default: (await import('./stories/aHeader')).AHeader }));
const Router = () => {
  if (document.location.search.indexOf('foo') !== -1) {
    return (
      <>
        <AHeader />
        <ZHeader />
      </>
    );
  }

  return <ZHeader />;
};

const root = ReactDOM.createRoot(document.getElementById('root') as HTMLElement);
root.render(
  <React.StrictMode>
    <Router />
  </React.StrictMode>
);
```

You can quite clearly see the same behaviour where the `ZHeader` renders green at `/?foo` and red at `/`:

![image](https://github.com/storybookjs/storybook/assets/132554/efd4e9fc-56d8-4a8a-a58b-952f7f27093d)
![image](https://github.com/storybookjs/storybook/assets/132554/73ed515b-5833-4996-95d0-2f54fb717cfb)

**Comment by hendrikheil** [comment]
So while we were working on an unrelated improvement for bundle size optimizations, we changed our upstream library to import each CSS module at the top of every ESM module we emit. This required quite a bit of tinkering with Vite, as it - by default - bundles all CSS and JS into two bundle files in Lib mode, which makes it a lot harder to tree-shake, but doesn't require processing by downstream bundlers.

By making that change, we now only rely on our global styles to provide certain reused design system tokens (as custom properties in :root), but that doesn't depend on any order in our stylesheets.

So for us, these changes made the regression go away, but I still believe storybook should behave the same way in dev and built mode - which it doesn't right now.

**Comment by tmeasday** [comment]
> So for us, these changes made the regression go away, but I still believe storybook should behave the same way in dev and built mode - which it doesn't right now.

Yes, this is a fair point. Do we have any idea of why it behaves differently? In the reproduction @martinkutter posted it broke in dev mode.

**Comment by hendrikheil** [comment]
Not sure really. I did see this issue upstream in vite: https://github.com/vitejs/vite/issues/3924 which might be related. I'm not sure where the root cause is really. I think it'd be interesting to see if the webpack builder exhibits the same behavior as the vite one.
